dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp

```diff
@@ -0,0 +1,166 @@
+
+#include "dawn/Optimizer/PassFixVersionedInputFields.h"
+#include "dawn-c/ErrorHandling.h"
+#include "dawn/AST/ASTExpr.h"
+#include "dawn/AST/ASTStmt.h"
+#include "dawn/IIR/ASTExpr.h"
+#include "dawn/IIR/ASTVisitor.h"
+#include "dawn/IIR/DoMethod.h"
+#include "dawn/IIR/IIRNodeIterator.h"
+#include "dawn/IIR/NodeUpdateType.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Support/Logging.h"
+
+#include <memory>
+
+namespace dawn {
+
+/// @brief Creates the assignment statement
+static std::shared_ptr<ast::Stmt>
+createAssignmentStatement(int assignmentID, int assigneeID,
+                          iir::StencilMetaInformation const& metadata) {
+  auto assignee = std::make_shared<ast::FieldAccessExpr>(metadata.getNameFromAccessID(assigneeID));
+  auto assignment =
+      std::make_shared<ast::FieldAccessExpr>(metadata.getNameFromAccessID(assignmentID));
+  auto assignmentExpression = std::make_shared<ast::AssignmentExpr>(assignment, assignee, "=");
+  auto assignmentStmt = iir::makeExprStmt(assignmentExpression);
+
+  // Add access IDs for the new access expressions
+  assignment->getData<iir::IIRAccessExprData>().AccessID = std::make_optional(assignmentID);
+  assignee->getData<iir::IIRAccessExprData>().AccessID = std::make_optional(assigneeID);
+
+  return assignmentStmt;
+}
+
+/// @brief Creates the do-method
+static std::unique_ptr<iir::DoMethod>
+createDoMethod(int assignmentID, int assigneeID,
+               std::shared_ptr<iir::StencilInstantiation> const& si,
+               const iir::Interval& interval) {
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+
+  // Create the assignment statement with the new and old variables
+  auto assignmentStmt = createAssignmentStatement(assignmentID, assigneeID, metadata);
+
+  iir::Accesses accesses;
+  accesses.addWriteExtent(assignmentID, iir::Extents{});
+  accesses.addReadExtent(assigneeID, iir::Extents{});
+  assignmentStmt->getData<iir::IIRStmtData>().CallerAccesses = std::move(accesses);
+
+  auto doMethod = std::make_unique<iir::DoMethod>(interval, metadata);
+
+  doMethod->getAST().push_back(std::move(assignmentStmt));
+  doMethod->update(iir::NodeUpdateType::level);
+
+  return doMethod;
+}
+
+/// @brief Creates the stage in which assignment happens (where the temporary gets filled)
+static std::unique_ptr<iir::Stage>
+createAssignmentStage(int assignmentID, int assigneeID,
+                      std::shared_ptr<iir::StencilInstantiation> const& si,
+                      const iir::Interval& interval) {
+  // Create the do-method for the stage
+  auto doMethod = createDoMethod(assignmentID, assigneeID, si, interval);
+
+  // Create the stage and add the do method
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+  auto assignmentStage = std::make_unique<iir::Stage>(metadata, si->nextUID());
+  assignmentStage->setExtents(iir::Extents{});
```

Comment:
If we do not want to run the metadata update pass, then we should get extents from the field instead of setting it to 0.